elastic/fleet-server
# Fleet-server package cache implementation

## What is the problem this PR solves?

If too many elastic-agents upgrade they may exhaust network resources.
By having fleet-server cache artifacts, it can assist with air-gapped deployments, and control the bandwidth used to serve artifacts. Coordination of agents will be done through other means.

## How does this PR solve the problem?

Allow the fleet-server to serve packages that the elastic-agent uses for
upgrades. The api endpoint is `GET /api/downloads/*path`. This endpoint
needs a valid API key. The endpoint for packages may be configured with
an associated concurrency limit and a bandwidth limit. The cache may have
an associated retention duration. When the fleet-server needs to retrieve a
package from the upstream source it will download the associated .sha512
and .asc files in order to verify the package. Upstream retrieval is a 1-1
mapping. i.e., `/api/downloads/some/artifact/package` will retrieve
`upstreamURI/downloads/some/artifact/package` validate it and store it on
the local filesystem as `cacheLocation/some/artifact/package`.

## How to test this PR locally

1. Start elasticsearch and kibana on the local machine
2. Create a service token and set the env var `ELASTICSEARCH_SERVICE_TOKEN`
3. Create a `fleet-server.yml` with
```yaml
output:
  elasticsearch:
    hosts: '${ELASTICSEARCH_HOSTS:localhost:9200}'
    service_token: '${ELASTICSEARCH_SERVICE_TOKEN}'

fleet:
  agent:
    id: 1e4954ce-af37-4731-9f4a-407b08e69e42
    logging:
      level: '${LOG_LEVEL:DEBUG}'

inputs:
  - type: fleet-server
      server:
        host: localhost
        port: 8220
      package_cache:
        enabled: false
        retention_period: -1
        upstreamURI: https://artifacts.elastic.co
        bandwidth_limit: -1
        concurrent_limit: -1

logging:
  to_stderr: true
http.enabled: true
```
4. Start a fleet server by running `./fleet-server -c fleet-server.yml`
5. Get an API Key from elasticsearch and it it to the env var `API_KEY`
6. Curl the fleet-server for an artifact: `curl  -H 'Authorization: ApiKey $API_KEY' localhost:8220/api/downloads/beats/elastic-agent/elastic-agent-8.1.2-linux-x86_64.tar.gz -o elastic-agent-8.1.2-linux-x86_64.tar.gz`
7. Check that the downloaded artifact is valid
8. Check the `fleet-package-cache` dir for the `.asc` and `.sha512` files

## Checklist





## Related issues
